rancilio-pid/rancilio-pid.ino

```diff
@@ -762,22 +774,45 @@ void brewdetection() {
 void ICACHE_RAM_ATTR onTimer1ISR() {
   timer1_write(50000); // set interrupt time to 10ms
 
-  if (Output <= isrCounter) {
-    digitalWrite(pinRelayHeater, LOW);
+  //run PID calculation
```

Comment:
die ISR muss so kurz wie möglich sein.
Ich bin  auch gegen das HeaterPreventFlapping, aus genau dem gleichen Grund wie oben.
Zusätzlich kommt hier hinzu, das diese logik nur aktiv ist wenn der Output < 100, oder > 900 ist.
Was außerhalb des normalen Regelbereichs liegt, und somit sogut wie nie zum Einsatz kommt.

Reply:
Nach meinen Tests sehe ich derzeit auch keine Sinn an diesem Feature. (auch wenn ich ein ungutes Gefühl habe den heater nur paar ms anzuhalten). Ich baue das Feature aus.

ABER in dem Code stecken auch Verbesserungen um die pid.compute() Berechnung an den isrCounter=0 zu "syncen". Das finde ich für wichtig, weil das Problemanalysen und zukünftige Tunings deutlich einfacher gestaltet. (Ich denke da mitunter an emergency fail-safes).

Reply:
Es ist aber nicht die Aufgabe einer ISR zu überwachen ob diese Sinnvoll verwendet wurde oder nicht.
Eine ISR die auf 1000 ms gestellt ist, wird auch alle 1000ms ausgeführt, egal ob der code darin ausgeführt wurde oder nicht.

Wenn wir das so implementieren wie du, dann ist die ISR nicht immer 1000ms lang, sondern kann auch mal länger sein, je nachdem wie "synchron" der PID das true zurück gibt, du hättest damit dann längere Heizzeiten, weil dein Zähler schon über 1000 ist.

Das ist nicht der Sinn dieser ISR

Reply:
Ich checke gleich mal eine um das beta feature bereinige version ein. Da lässt sich der Unterschied gut sehen. 

Die Begründung mit längeren Heizzeiten kann ich nicht nachvollziehen. 
Hingegen bin ich sehr sicher, dass wir stattdessen unnötige An/Aus-Schaltungen der Heizung haben: Stell dir vor was beispielsweise passiert, wenn 
* der compute() konstant um 800ms startet
* compute() zufälligerweise output Regelungen in der Nähe dieses Zeitpunkts durchführt (dh output=+/-800).
zB 750, 850,750: In derzeitigen ISR-Code zähle ich 8 An/Aus-Aktionen. Im neuen Code nur 6.

Ich kann mir nicht vorstellen dass die Hardware es gerne hat, wenn sie innerhalb weniger millisekunden Schaltoperationen durchführen muss. Und selbst wenn sie das aushält, dann ist es nichts was ich gerne sehe.
Weiterer Vorteil ist dass das debugging/analyse viel einfacher ist.

Reply:
@Miau09 schau dir den Code und meine Beschreibung bitte an. Ich habe nun zum Vergleich den alten und neuen ISR code drinnen.